File: qspray/inst/keep/BBdivision.cpp
// Rcpp::List BBdivisionRcpp(
//   Rcpp::List Powers, Rcpp::StringVector coeffs, Rcpp::List LTsf, 
//   Rcpp::List gs, Rcpp::List LTgs, int d
// ) {
//   int ngs    = gs.size();
//   int nterms = LTsf.size(); 
//   qspray cur = makeQspray(Powers, coeffs);
//   for(int k = 0; k < nterms; k++) {
//     Rcpp::List LTcur = LTsf(k);
//     int i = 0;
//     while(i < ngs) {
//       Rcpp::List g  = gs(i);
//       qspray gspray = makeQspray(g["powers"], g["coeffs"]);
//       Rcpp::List LTg = LTgs(i);
//       while(divides(LTg, LTcur)) {
//         qspray qtnt = quotient(LTcur, LTg);
//         cur = subtract(cur, prod(qtnt, gspray));
//         if(cur.size() == 0) {
//           return(retval(cur));
//         }
//         LTcur = leadingTerm(cur, d);
//       }
//       i++;
//     }
//   }
//   return retval(cur);
// }

File: qspray/inst/keep/BBdivision.R
# internal division for Buchberger algorithm
BBdivision <- function(qspray, divisors, LTdivisors) {
  if(qspray == qzero()) {
    return(qzero())
  }
  # we store the successive leading terms in LTs_f
  d <- max(arity(qspray), max(vapply(divisors, arity, integer(1L))))
  oqspray <- orderedQspray(qspray, d)
  opowers <- oqspray[["powers"]]
  ocoeffs <- oqspray[["coeffs"]]
  LTs_f <- lapply(seq_along(ocoeffs), function(i) {
    list("powers" = opowers[i, ], "coeff" = ocoeffs[i])
  })
  
  ndivisors <- length(divisors)
  nterms <- length(qspray@coeffs)

  cur <- qspray
  for(k in 1L:nterms) {
    LT_cur <- LTs_f[[k]]
    i <- 1L
    while(i <= ndivisors) {
      g    <- divisors[[i]]
      LT_g <- LTdivisors[[i]] 
      while(divides(LT_g, LT_cur)) {
        cur <- cur - quotient(LT_cur, LT_g) * g
        if(cur == qzero()) {
          return(qzero())
        }
        LT_cur <- leadingTerm(cur, d)
      }
      i <- i + 1L
    }
  }
  # return remainder
  cur
}
